File: Lab1/lista_dwukier.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
	int value;
	struct node* next;
	struct node* prev;
} node;

node* newNode(int data) {
	node* tmp = malloc(sizeof(node));
	tmp->value = data;
	tmp->next = NULL;
	tmp->prev = NULL;
	return tmp;
}

void push_front(node **head, int data) {
	node* tmp = *head;
	node* New = newNode(data);
	if (tmp == NULL) {
		*head = New;
		return head;
	}

	tmp->prev = New;
	New->next = tmp;
	*head = New;
}

void push_back(node* head, int data) {
	node* tmp = head;
	node* New = newNode(data);

	while (tmp->next != NULL) {
		tmp = tmp->next;
	}
	New->prev = tmp;
	tmp->next = New;
}

int deleteIndex(node** head, int index) {
	node *tmp = *head;
	int val = tmp->value;
	if (index == 0 && tmp->next == NULL)
		free(tmp);
	else if(index == 0 && tmp->next != NULL) {
		tmp->next->prev = NULL;
		*head = tmp->next;
		return val;
	}

	for (int i = 0; i < index; i++)
	{
		tmp = tmp->next;
		if (tmp == NULL) {
			printf("index out of range exception\n");
			return -216521;
		}
	}

	tmp->prev->next = tmp->next;
	if (tmp->next != NULL)
		tmp->next->prev = tmp->prev;
	free(tmp);
	return val;

}

void showAll(node* head) {
	node* tmp = head;
	do {
		printf("%d ", tmp->value);
		tmp = tmp->next;
	} while (tmp != NULL);
	printf("\n");
}

main() {

	node* head = newNode(1);
	push_back(head, 1);
	push_back(head, 2);
	push_back(head, 3);

	push_front(&head, 0);
	showAll(head);

	deleteIndex(&head, 0);
	showAll(head);
	deleteIndex(&head, 2);
	showAll(head);
	deleteIndex(&head, 2);
	showAll(head);
	deleteIndex(&head, 10);
	return 0;
}

File: Lab1/lista_dwukier.py
class Node:
    def __init__(self, val):
        self.value = val
        self.next = None
        self.prev = None

class List:
    def __init__(self):
        self.head = None
    
    def push_back(self, data):
        New = Node(data)
        if self.head is None:
            self.head = New
        else:
            tmp = self.head
            while tmp.next is not None:
                tmp = tmp.next
            tmp.next = New
            New.prev = tmp

    def push_front(self, data):
        New = Node(data)
        if self.head is None:
            self.head = New
        else:
            tmp = self.head
            tmp.prev = New
            New.next = tmp
            self.head = New

    def showAll(self):
        tmp = self.head
        while tmp is not None:
            print("{} ".format(tmp.value), end="")
            tmp = tmp.next
        print("")

    def deleteIndex(self, index):
        if index == 0 and self.head.next is None:
            self.head = None
            return
        elif index == 0 and self.head.next is not None:
            self.head.next.prev = None
            self.head = self.head.next
            return

        tmp = self.head
        for i in range(index):
            tmp = tmp.next
            if tmp is None:
                print("index out of range exception")
                return
        if tmp.next is not None:
            tmp.prev.next = tmp.next
            tmp.next.prev = tmp.prev
            tmp = None
        else:
            tmp.prev.next = None
            tmp = None



lista = List()
lista.push_back(1)
lista.push_back(2)
lista.push_back(3)
lista.push_front(0)
lista.showAll()
lista.deleteIndex(0)
lista.showAll()
lista.deleteIndex(1)
lista.showAll()
lista.push_back(4)
lista.showAll()
lista.deleteIndex(2)
lista.showAll()
lista.deleteIndex(5)
